File: swim-system-java/swim-core-java/swim.uri/src/main/java/swim/uri/UriPort.java
// Copyright 2015-2021 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.uri;

import swim.codec.Debug;
import swim.codec.Display;
import swim.codec.Format;
import swim.codec.Output;
import swim.util.HashGenCacheSet;
import swim.util.Murmur3;

public class UriPort implements Comparable<UriPort>, Debug, Display {

  protected final int number;

  protected UriPort(int number) {
    this.number = number;
  }

  public final boolean isDefined() {
    return this.number != 0;
  }

  public final int number() {
    return this.number;
  }

  @Override
  public int compareTo(UriPort that) {
    return Integer.compare(this.number, that.number);
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) {
      return true;
    } else if (other instanceof UriPort) {
      return this.number == ((UriPort) other).number;
    }
    return false;
  }

  private static int hashSeed;

  @Override
  public int hashCode() {
    if (UriPort.hashSeed == 0) {
      UriPort.hashSeed = Murmur3.seed(UriPort.class);
    }
    return Murmur3.mash(Murmur3.mix(UriPort.hashSeed, this.number));
  }

  @Override
  public <T> Output<T> debug(Output<T> output) {
    output = output.write("UriPort").write('.');
    if (this.isDefined()) {
      output = output.write("create").write('(');
      output = Format.displayInt(this.number, output);
      output = output.write(')');
    } else {
      output = output.write("undefined").write('(').write(')');
    }
    return output;
  }

  @Override
  public <T> Output<T> display(Output<T> output) {
    return Format.displayInt(this.number, output);
  }

  @Override
  public String toString() {
    return Integer.toString(this.number);
  }

  private static UriPort undefined;

  public static UriPort undefined() {
    if (UriPort.undefined == null) {
      UriPort.undefined = new UriPort(0);
    }
    return UriPort.undefined;
  }

  public static UriPort create(int number) {
    if (number > 0) {
      return UriPort.cache().put(new UriPort(number));
    } else if (number == 0) {
      return UriPort.undefined();
    } else {
      throw new IllegalArgumentException(Integer.toString(number));
    }
  }

  public static UriPort parse(String string) {
    return Uri.standardParser().parsePortString(string);
  }

  private static ThreadLocal<HashGenCacheSet<UriPort>> cache = new ThreadLocal<>();

  static HashGenCacheSet<UriPort> cache() {
    HashGenCacheSet<UriPort> cache = UriPort.cache.get();
    if (cache == null) {
      int cacheSize;
      try {
        cacheSize = Integer.parseInt(System.getProperty("swim.uri.port.cache.size"));
      } catch (NumberFormatException e) {
        cacheSize = 16;
      }
      cache = new HashGenCacheSet<UriPort>(cacheSize);
      UriPort.cache.set(cache);
    }
    return cache;
  }

}

File: swim-system-js/swim-core-js/@swim/uri/main/UriPort.ts
// Copyright 2015-2020 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Comparable, HashCode, Murmur3, HashGenCacheMap} from "@swim/util";
import {Output, Format, Debug, Display} from "@swim/codec";
import {Uri} from "./Uri";

export type AnyUriPort = UriPort | number | string;

export class UriPort implements Comparable<UriPort>, HashCode, Debug, Display {
  /** @hidden */
  readonly _number: number;

  /** @hidden */
  constructor(num: number) {
    this._number = num;
  }

  isDefined(): boolean {
    return this._number !== 0;
  }

  number(): number {
    return this._number;
  }

  valueOf(): number {
    return this._number;
  }

  toAny(): number {
    return this._number;
  }

  compareTo(that: UriPort): 0 | 1 | -1 {
    return this._number < that._number ? -1 : this._number > that._number ? 1 : 0;
  }

  equals(that: unknown): boolean {
    if (this === that) {
      return true;
    } else if (that instanceof UriPort) {
      return this._number === that._number;
    }
    return false;
  }

  hashCode(): number {
    if (UriPort._hashSeed === void 0) {
      UriPort._hashSeed = Murmur3.seed(UriPort);
    }
    return Murmur3.mash(Murmur3.mix(UriPort._hashSeed, this._number));
  }

  debug(output: Output): void {
    output = output.write("UriPort").write(46/*'.'*/);
    if (this.isDefined()) {
      output = output.write("from").write(40/*'('*/);
      Format.displayNumber(this._number, output);
      output = output.write(41/*')'*/);
    } else {
      output = output.write("undefined").write(40/*'('*/).write(41/*')'*/);
    }
  }

  display(output: Output): void {
    Format.displayNumber(this._number, output);
  }

  toString(): string {
    return "" + this._number;
  }

  private static _hashSeed?: number;

  private static _undefined?: UriPort;

  private static _cache?: HashGenCacheMap<number, UriPort>;

  static undefined(): UriPort {
    if (UriPort._undefined === void 0) {
      UriPort._undefined = new UriPort(0);
    }
    return UriPort._undefined;
  }

  static from(number: number) {
    if (number > 0) {
      const cache = UriPort.cache();
      const port = cache.get(number);
      if (port !== void 0) {
        return port;
      } else {
        return cache.put(number, new UriPort(number));
      }
    } else if (number === 0) {
      return UriPort.undefined();
    } else {
      throw new TypeError("" + number);
    }
  }

  static fromAny(port: AnyUriPort | null | undefined): UriPort {
    if (port === null || port === void 0) {
      return UriPort.undefined();
    } else if (port instanceof UriPort) {
      return port;
    } else if (typeof port === "number") {
      return UriPort.from(port);
    } else if (typeof port === "string") {
      return UriPort.parse(port);
    } else {
      throw new TypeError("" + port);
    }
  }

  static parse(string: string): UriPort {
    return Uri.standardParser().parsePortString(string);
  }

  /** @hidden */
  static cache(): HashGenCacheMap<number, UriPort> {
    if (UriPort._cache === void 0) {
      const cacheSize = 4;
      UriPort._cache = new HashGenCacheMap<number, UriPort>(cacheSize);
    }
    return UriPort._cache;
  }
}
Uri.Port = UriPort;
